ssl/t1_lib.c

```diff
@@ -4509,14 +4524,28 @@ static const SIGALG_LOOKUP *find_sig_alg(SSL_CONNECTION *s, X509 *x,
 
     /* Look for a shared sigalgs matching possible certificates */
     for (i = 0; i < s->shared_sigalgslen; i++) {
-        lu = s->shared_sigalgs[i];
+        int dtls, minversion, maxversion;
 
         /* Skip SHA1, SHA224, DSA and RSA if not PSS */
+        lu = s->shared_sigalgs[i];
         if (lu->hash == NID_sha1
             || lu->hash == NID_sha224
             || lu->sig == EVP_PKEY_DSA
             || lu->sig == EVP_PKEY_RSA)
             continue;
+
+        /*
+         * By this point the protocol version should already be chosen.  Check
+         * the sigalg version bounds.
+         */
+        dtls = SSL_CONNECTION_IS_DTLS(s);
+        minversion = dtls ? lu->mindtls : lu->mintls;
+        maxversion = dtls ? lu->maxdtls : lu->maxtls;
+        if (minversion != 0 && ssl_version_cmp(s, minversion, s->version) > 0)
+            continue;
+        if (maxversion != 0 && ssl_version_cmp(s, maxversion, s->version) < 0)
+            continue;
```

Comment:
Should we also be calling `tls12_sigalg_allowed`?

Reply:
The `lu` objects considered are already accumulated in `s->shared_sigalgs[i]`, I was assuming that the check would have aready happened when that intersection was built, but it is worth checking that this is true...

This appears to be done in `tls1_set_shared_sigalgs()->tls12_shared_sigalgs()`, the latter does call `tls12_sigalg_allowed()`.  It does not have version checks, but we do that perhaps in too many places already, and sometimes before we know our version, so perhaps good enough?